modules/library/render/src/main/java/org/geotools/renderer/label/LabelCacheImpl.java

```diff
@@ -571,6 +591,369 @@ private double goodnessOfFit(LabelPainter painter, AffineTransform transform,
+            reverse.moveTo(cursor.getLineStringLength() - cursor.getCurrentOrdinate());
+            cursor = reverse;
+        }
+
+        // find out the true centering position
+        double anchorY = painter.getLinePlacementYAnchor();
+
+        // init, move to the starting position
+        double mid = cursor.getCurrentOrdinate();
+        Coordinate c = new Coordinate();
+        c = cursor.getCurrentPosition(c);
+
+        double startOrdinate = mid - painter.getStraightLabelWidth() / 2;
+        if (startOrdinate < 0)
+            startOrdinate = 0;
+        cursor.moveTo(startOrdinate);
+
+        for (LineInfo.LineComponent lineComponent : lineInfo.getComponents()) {
+            GlyphVector gv = lineComponent.getGlyphVector();
+            float nextAdvance = gv.getGlyphMetrics(0).getAdvance() * 0.5f;
+            for (int i = 0; i < gv.getNumGlyphs(); i++) {
+                Point2D p = gv.getGlyphPosition(i);
+                float advance = nextAdvance;
+                nextAdvance = i < gv.getNumGlyphs() - 1
+                        ? gv.getGlyphMetrics(i + 1).getAdvance() * 0.5f
+                        : 0;
+                c = cursor.getCurrentPosition(c);
+                AffineTransform t = new AffineTransform();
+                t.setToTranslation(c.x, c.y);
+                t.rotate(cursor.getCurrentAngle());
+                t.translate(-p.getX() - advance, -p.getY() + painter.getLineHeight() * anchorY);
+                transforms.add(t);
+
+                cursor.moveTo(cursor.getCurrentOrdinate() + advance + nextAdvance);
+            }
+        }
+        return transforms;
+    }
+
+    /**
+     * Compute the affine transform for each glyph along a straight line.
+     * Does not consider letter orientation (only letter position).
+     * (taken from LabelPainter#paintCurvedLabel).
+     * @param painter the painter in charge of painting the glyphs
+     * @param tx global label AffineTranform
+     * @return a list of AffinTransforms (one AffineTransform per glyph)
+     */
+    private List<AffineTransform> getGlyphTransforms(LabelPainter painter, AffineTransform tx){
+        List<AffineTransform> transforms = new ArrayList<>();
+        for (LineInfo lineInfo : painter.lines) {
+            for (LineInfo.LineComponent lineComponent : lineInfo.getComponents()) {
+                AffineTransform componentTx = new AffineTransform(tx);
+                componentTx.translate(lineComponent.getX(), lineInfo.getY());
+                for (int i = 0; i < lineComponent.getGlyphVector().getNumGlyphs(); i++) {
+                    transforms.add(componentTx);
+                }
+            }
+        }
+        return transforms;
+    }
+
+    //Modified version of paintLineLabels
+    //We compute the Bounding box for each letters instead of the whole label
+    //then we check each letters for collisions
+    private boolean paintLineLabelsWithLetterConflict(LabelPainter painter, AffineTransform originalTransform,
```

Comment:
Hi,

I refactored this part of the code with a GlyphVectorProcessor class which has two subclasses :
- one to process Straight line labels
- one to process Curved line labels
Their process method can take a GlyphProcessor instance which can do one of the following (depending on the subclass) :
- draw the letter bounding box
- test if the letter conflicts
- put the letter in one index or the other

If this refactored code is approved, it could be interesting to see how it can be shared with LabelPainter as a part of the code comes from the LabelPainter class...

Michaël
________________________________
De : Andrea Aime [notifications@github.com]
Date d'envoi : jeudi 27 août 2015 08:36
À : geotools/geotools
Cc : Michael Michaud
Objet : Re: [geotools] LineLabel placement improved (#948)


In modules/library/render/src/main/java/org/geotools/renderer/label/LabelCacheImpl.java<https://github.com/geotools/geotools/pull/948#discussion_r38066090>:

> +        for (LineInfo lineInfo : painter.lines) {
> +            for (LineInfo.LineComponent lineComponent : lineInfo.getComponents()) {
> +                AffineTransform componentTx = new AffineTransform(tx);
> +                componentTx.translate(lineComponent.getX(), lineInfo.getY());
> +                for (int i = 0; i < lineComponent.getGlyphVector().getNumGlyphs(); i++) {
> +                    transforms.add(componentTx);
> +                }
> +            }
> +        }
> +        return transforms;
> +    }
> +
> +    //Modified version of paintLineLabels
> +    //We compute the Bounding box for each letters instead of the whole label
> +    //then we check each letters for collisions
> +    private boolean paintLineLabelsWithLetterConflict(LabelPainter painter, AffineTransform originalTransform,


I'm not comfortable with the amount of code duplication happening here, most of the method is identical to paintLineLabels, the code that's handling the conflicts should be factored out, could be a separate little class with two subclasses doing the checks and updating the confilct quadtree.

If you don't manage to do it by tomorrow I understand and we can merge this now, but under the condition that the code duplication will be fixed sometimes soon, e.g., the next week or the week after.

—
Reply to this email directly or view it on GitHub<https://github.com/geotools/geotools/pull/948/files#r38066090>.